snooty/postprocess.py

```diff
@@ -919,6 +919,57 @@ def enter_node(self, fileid_stack: FileIdStack, node: n.Node) -> None:
             self.guides[current_slug].description = node.children
 
 
+class OpenAPIHandler(Handler):
+    """Constructs metadata for OpenAPI content pages."""
+
+    @dataclass
+    class SourceData:
+        source_type: str
+        source: str
+
+    def __init__(self, context: Context) -> None:
+        super().__init__(context)
+        self.openapi_pages: Dict[str, OpenAPIHandler.SourceData] = {}
+
+    def get_metadata(self) -> Dict[str, SerializableType]:
+        """Returns serialized object to be used as part of the build's metadata."""
+
+        return {k: asdict(v) for k, v in self.openapi_pages.items()}
+
+    def enter_node(self, fileid_stack: FileIdStack, node: n.Node) -> None:
+        if (
+            not isinstance(node, n.Directive)
+            or node.name != "openapi"
+            or node.options.get("preview")
+        ):
+            return
+
+        current_file = fileid_stack.current
+        current_slug = clean_slug(current_file.without_known_suffix)
+
+        if current_slug in self.openapi_pages:
+            self.context.diagnostics[current_file].append(
+                DuplicateDirective(node.name, node.start[0])
+            )
+            return
+
+        # source_type should be assigned in the parsing layer
+        source_type = node.options.get("source_type")
+        if not source_type:
+            return
+
+        source = ""
+        argument = node.argument[0]
+        if source_type == "local" or source_type == "atlas":
+            assert isinstance(argument, n.Text)
+            source = argument.get_text()
+        else:
+            assert isinstance(argument, n.Reference)
```

Comment:
is this guaranteed to be true? i.e. can a writer subvert this?

Reply:
The comment for the line above should also answer this same question! Let me know if it doesn't though